[mlir][vector] Add mask elimination transform

MacDue opened the pull request:
This adds a new transform `eliminateVectorMasks()` which aims at removing scalable `vector.create_masks` that will be all-true at runtime. It attempts to do this by simply pattern-matching the mask operands (similar to some canonicalizations), if that does not lead to an answer (is all-true? yes/no), then value bounds analysis will be used to find the lower bound of the unknown operands. If the lower bound is >= to the corresponding mask vector type dim, then that dimension of the mask is all true.

Note that the pattern matching prevents expensive value-bounds analysis in cases where the mask won't be all true.

For example:
```mlir
%mask = vector.create_mask %dynamicValue, %c2 : vector<8x4xi1>
```
From looking at `%c2` we can tell this is not going to be an all-true mask, so we don't need to run the value-bounds analysis for
`%dynamicValue` (and can exit the transform early).

Note: Eliminating create_masks here means replacing them with all-true constants (which will then lead to the masks folding away).

MacDue committed:
[mlir][vector] Add mask elimination transform

This adds a new transform `eliminateVectorMasks()` which aims at
removing scalable `vector.create_masks` that will be all-true at
runtime. It attempts to do this by simply pattern-matching the mask
operands (similar to some canonicalizations), if that does not lead to
an answer (is all-true? yes/no), then value bounds analysis will be used
to find the lower bound of the unknown operands. If the lower bound is
>= to the corresponding mask vector type dim, then that dimension of the
mask is all true.

Note: Eliminating create_masks here means replacing them with all-true
constants (which will then lead to the masks folding away).

hanhanW commented on mlir/lib/Dialect/Vector/Transforms/VectorMaskElimination.cpp:
I think you miss the license. E.g.,

```
//===- ................... --*- C++ -*-===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//
```

c-rhodes commented on mlir/test/Dialect/Vector/eliminate-masks.mlir:
```diff
@@ -0,0 +1,138 @@
+    %mask = vector.create_mask %dim : vector<[4]xi1>
+    "test.some_use"(%mask) : (vector<[4]xi1>) -> ()
+    // !!! Here the size of the mask could shrink in the next iteration.
+    %next_num_els = affine.min  affine_map<(d0)[s0] -> (-d0 + 1000, s0)>(%i)[%c4_vscale]
```
```suggestion
    %next_num_elts = affine.min  affine_map<(d0)[s0] -> (-d0 + 1000, s0)>(%i)[%c4_vscale]
```

c-rhodes commented on mlir/test/Dialect/Vector/eliminate-masks.mlir:
```diff
@@ -0,0 +1,138 @@
+  %vscale = vector.vscale
+  %c4_vscale = arith.muli %vscale, %c4 : index
+  %extracted_slice_0 = tensor.extract_slice %tensor[0, 0] [1, %c4_vscale] [1, 1] : tensor<1x1000xf32> to tensor<1x?xf32>
+  %output_tensor = scf.for %i = %c0 to %c1000 step %c4_vscale iter_args(%arg = %extracted_slice_0) -> tensor<1x?xf32> {
```
this loop does the same thing after the first iteration (induction var isn't used), is that intentional?

MacDue replied:
not really :+1:

c-rhodes commented on mlir/test/Dialect/Vector/eliminate-masks.mlir:
```diff
@@ -0,0 +1,138 @@
+  %c4_vscale = arith.muli %vscale, %c4 : index
+  %extracted_slice = tensor.extract_slice %tensor[0] [%c4_vscale] [1] : tensor<1000xf32> to tensor<?xf32>
+  %slice = scf.for %i = %c0 to %c1000 step %c4_vscale iter_args(%arg = %extracted_slice) -> tensor<?xf32> {
+    // This mask cannot be eliminated even though looking at the above operations
+    // it appears `tensor.dim` will always be c4_vscale (so the mask all-true).
```
but `tensor.dim` doesn't always appear to be `%c4_vscale`, that would only be considering the initial value?

c-rhodes commented on mlir/test/Dialect/Vector/eliminate-masks.mlir:
```diff
@@ -0,0 +1,138 @@
+  %c4 = arith.constant 4 : index
+  %vscale = vector.vscale
+  %c4_vscale = arith.muli %vscale, %c4 : index
+  // This is _very_ simple but since addi is not a constant value bounds will
```
```suggestion
  // This is _very_ simple but since muli is not a constant value bounds will
```

MacDue replied:
That's meant to say `tensor.dim`

c-rhodes commented on mlir/test/lib/Dialect/Vector/TestVectorTransforms.cpp:
```diff
@@ -874,6 +874,38 @@ struct TestVectorLinearize final
+  Option<unsigned> vscaleMin{
+      *this, "vscale-min",
+      llvm::cl::desc(
+          "Minimum value `vector.vscale` can possibly be at runtime."),
```
nit:
```suggestion
          "Minimum value of `vector.vscale` at runtime."),
```

MacDue replied:
This has a different meaning to me. 
i.e. the minimum value cross all possible runtimes vs the minimum value of vscale for a specific runtime.

c-rhodes replied:
not sure I follow, but I don't think it necessary to mention runtime anyway, "Minimum value of `vscale.vscale`" is sufficient.

MacDue replied:
I've shortened it to "Minimum possible value of vscale." (which is used elsewhere too)

c-rhodes commented on mlir/test/lib/Dialect/Vector/TestVectorTransforms.cpp:
```diff
@@ -874,6 +874,38 @@ struct TestVectorLinearize final
+  Option<unsigned> vscaleMax{
+      *this, "vscale-max",
+      llvm::cl::desc(
+          "Maximum value `vector.vscale` can possibly be at runtime."),
```
nit:
```suggestion
          "Maximum value of `vector.vscale` at runtime."),
```

c-rhodes commented on mlir/test/lib/Dialect/Vector/TestVectorTransforms.cpp:
```diff
@@ -874,6 +874,38 @@ struct TestVectorLinearize final
+      *this, "vscale-max",
+      llvm::cl::desc(
+          "Maximum value `vector.vscale` can possibly be at runtime."),
+      llvm::cl::init(16)};
```
this is specific to SVE, please add a TODO to factor this out into target info.

MacDue replied:
This is just a test pass so it probably won't ever have/use target info.

MacDue commented on mlir/lib/Dialect/Vector/Transforms/VectorMaskElimination.cpp:
```diff
@@ -0,0 +1,117 @@
+    return;
+
+  OpBuilder::InsertionGuard g(rewriter);
+  SmallVector<vector::CreateMaskOp> worklist;
+  function.walk([&](vector::CreateMaskOp createMaskOp) {
+    worklist.push_back(createMaskOp);
+  });
+  rewriter.setInsertionPointToStart(&function.front());
+  for (auto mask : worklist)
+    (void)resolveAllTrueCreateMaskOp(rewriter, mask, *vscaleRange);
```
It's not safe to mutate and walk IR at the same time.

joker-eph replied:
This statement as a generality does not seem correct: a lot of mutation are perfectly fine to do while walking the IR, it all depends what you do. 
I would say that if you need a worklist, you should document why the delayed modification is needed.

MacDue replied:
I would think modifying existing ops is okay, but inserting new operations could invalidate the walker?

MacDue commented on mlir/lib/Dialect/Vector/Transforms/VectorMaskElimination.cpp:
```diff
@@ -0,0 +1,117 @@
+            dimSize, {}, vscaleRange.vscaleMin, vscaleRange.vscaleMax,
+            presburger::BoundType::LB);
+    if (failed(lowerBound))
+      return failure();
```
This is just a function not a pattern, so I don't think `notifyMatchFailure()` would do anything here.

MacDue committed:
Fixups

dcaballe commented:
Thanks! A couple of high-level comments. I'll come back :)

1. Could we move the trivial pattern-matching cases to the folder/canonicalizer? We really don't want to have any trivial all-one mask around both for the fixed and scalable cases. It makes sense to have a more sophisticated and expensive pass to get rid of the tricky cases that require value bounds or range analysis.

2. Could we use `constant_mask` instead of `arith.constant`? This is not enforced in any way but we've been using the former to represent vector masks (they are kind of redundant, actually).

MacDue commented:
1. The simple pattern matching is done while collecting the unknown dimensions, and functions as an early-exit, as if we can tell that a mask is not going to be all-true we can bail out before doing any more expensive value-bounds analysis. Patterns like this already exist as canonizations, but just give-up on unknown dims rather than trying to resolve them. 

https://github.com/llvm/llvm-project/blob/7af1229b9f2cfab5d5437bf34bbbb272dd9108b6/mlir/lib/Dialect/Vector/Transforms/VectorMaskElimination.cpp#L50-L66

2. I think I'd need to add a nice builder for making all-true/false constant masks first, currently, it's much less ergonomic to build a constant_mask vs an arith.constant, since the only builder for constant_masks takes an ArrayAttr... So you have to build an array of IntegerAttr, which is a bit of a pain.

banach-space commented on mlir/test/Dialect/Vector/eliminate-masks.mlir:
```diff
@@ -0,0 +1,138 @@
+  %c4_vscale = arith.muli %vscale, %c4 : index
+  %extracted_slice = tensor.extract_slice %tensor[0] [%c4_vscale] [1] : tensor<1000xf32> to tensor<?xf32>
+  %slice = scf.for %i = %c0 to %c1000 step %c4_vscale iter_args(%arg = %extracted_slice) -> tensor<?xf32> {
+    // This mask cannot be eliminated even though looking at the above operations
+    // it appears `tensor.dim` will always be c4_vscale (so the mask all-true).
```
[nit] Might be just me, but I just find it clearer when you say specifically "Ops above this particular Op" as opposed to "above Ops" (in general, some set of Ops above).
```suggestion
    // This mask cannot be eliminated even though looking at the operations
    // above `tensor.dim`, it appears it will always be c4_vscale (so the mask all-true).
```

MacDue replied:
I changed it to "above this comment" as in the latter half of the sentence I'm referring to the value of `tensor.dim` "it appears `tensor.dim` will always be c4_vscale", changing that to "it" is makes it less clear to me.

banach-space commented on mlir/test/Dialect/Vector/eliminate-masks.mlir:
```diff
@@ -0,0 +1,138 @@
+// CHECK-NOT: arith.constant dense<true> : vector<2x[4]xi1>
+// CHECK: %[[MASK:.*]] = vector.create_mask
+// CHECK: "test.some_use"(%[[MASK]]) : (vector<2x[4]xi1>) -> ()
+func.func @negative_constant_dim_not_all_true()
```
[nit] Similar comment below - just to better contrast the two.
```suggestion
func.func @negative_fixed_width_dim_not_all_true()
```

MacDue replied:
This is testing a constant dim in `vector.create_mask` not being all-true, not really the fixed-with part.

MacDue replied:
I've added an extra comment inside this test case to explain it.

banach-space commented on mlir/test/Dialect/Vector/eliminate-masks.mlir:
```diff
@@ -0,0 +1,138 @@
+// CHECK-NOT: arith.constant dense<true> : vector<2x[4]xi1>
+// CHECK: %[[MASK:.*]] = vector.create_mask
+// CHECK: "test.some_use"(%[[MASK]]) : (vector<2x[4]xi1>) -> ()
+func.func @negative_constant_vscale_multiple_not_all_true() {
```
[nit] Similar comment the example above - just to better contrast the two.
```suggestion
func.func @negative_scalabie_dim_not_all_true() {
```

MacDue replied:
This is testing a constant vscale multiple (3 * vscale), not being all-true, not really the scalable part.

MacDue replied:
I've added an extra comment inside this test case to explain it.

banach-space commented on mlir/test/Dialect/Vector/eliminate-masks.mlir:
```diff
@@ -0,0 +1,138 @@
+  %c4 = arith.constant 4 : index
+  %vscale = vector.vscale
+  %c4_vscale = arith.muli %vscale, %c4 : index
+  // This is _very_ simple but since tensor.dim is not a constant value bounds
+  // will be used to resolve it.
```
I don't follow this comment (emphasis mine):
> since tensor.dim is not a constant value bounds **will be used** to **resolve** it

IIUC, all tests in this file rely on "value bounds analysis". As in, t is is always used - why call it out here specifically? Is there anything special in this test? 

Also, "value bounds analysis" will fail in this case because the dimension size is not know at compile time (so it cannot be inferred), right?

MacDue replied:
1. Not all tests in this file will use value bounds, the simple cases (tested before these), will be found to be not all-true via simple pattern matching. 

2. The dim size is 2 so value bounds will easily find it.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorMaskElimination.cpp:
```diff
@@ -0,0 +1,129 @@
+    return {};
+  auto lhs = mul.getLhs();
+  auto rhs = mul.getRhs();
+  if (lhs.getDefiningOp<vector::VectorScaleOp>())
+    return getConstantIntValue(rhs);
+  if (rhs.getDefiningOp<vector::VectorScaleOp>())
+    return getConstantIntValue(lhs);
```
IIUC, this assumes one of 3 cases:
* `int * int`
* `vscale * int`
* `int * vscale`

How about `vscale * vscale`? We should capture and test that as well.

MacDue replied:
I'm not sure how it does? :sweat_smile: 

`int * int` is not a valid case nor is `vscale * vscale`.

MacDue replied:
That's also `std::nullopt` yes :)

banach-space replied:
Yeah, that wasn't obvious to me when I first read it.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorMaskElimination.cpp:
```diff
@@ -0,0 +1,129 @@
+  for (auto [i, dimSize] : llvm::enumerate(createMaskOp.getOperands())) {
+    if (auto intSize = getConstantIntValue(dimSize)) {
+      // Mask not all-true for this dim.
+      if (maskTypeDimScalableFlags[i] || intSize < maskTypeDimSizes[i])
+        return failure();
```
IIUC, `getConstantIntValue` would return `nullopt` for `vscale`? Shouldn't it always be failure if the mask is scalable and `intSize` is _not_ a multiple of `vscale` (i.e. a compile-time known int).

MacDue replied:
It is a failure if the mask is scalable here. If `maskTypeDimScalableFlags[i]` then it returns failure.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorMaskElimination.cpp:
```diff
@@ -0,0 +1,129 @@
+  for (auto [i, dimSize] : unknownDims) {
+    // Compute the lower bound for the unknown dimension (i.e. the smallest
+    // value it could be).
+    auto lowerBound =
```
1. Spell out `auto`.
2. `lowerBound` of what?

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorMaskElimination.cpp:
```diff
@@ -0,0 +1,129 @@
+            presburger::BoundType::LB);
+    if (failed(lowerBound))
+      return failure();
+    auto boundSize = lowerBound->getSize();
```
```suggestion
    auto lowerBoundSize = lowerBound->getSize();
```
Let's avoid generic names where a very specific value is requested.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorMaskElimination.cpp:
```diff
@@ -0,0 +1,129 @@
+    if (failed(boundSize))
+      return failure();
+    if (boundSize->scalable) {
+      // If the lower bound is scalable and >= to the mask dim size then this
+      // dim is all-true.
+      if (boundSize->baseSize < maskTypeDimSizes[i])
+        return failure();
```
This is a bit confusing. Surely we want the mask dim to be >= tensor dim for ALL TRUE, right?
```
// NOT ALL TRUE (%c2 < 4) 
%0 = vector.create_mask %c2, %c3 : vector<4x3xi1>
// ALL TRUE (%c5 >= 4)
%0 = vector.create_mask %c5, %c3 : vector<4x3xi1>
```

Btw, it would help if the condition that's checked (`<`) and the comment (`>=`) were consistent.

MacDue replied:
Updated comment.

MacDue committed:
Comments

dcaballe commented:
> The simple pattern matching is done while collecting the unknown dimensions, and functions as an early-exit, as if we can tell that a mask is not going to be all-true we can bail out before doing any more expensive value-bounds analysis. Patterns like this already exist as canonizations, but just give-up on unknown dims rather than trying to resolve them

Sorry, I'm not sure what this comment implies but, yes, I'm meant that the extra pattern matching should be part of canonicalization. If you also need those patterns here to make the pass reach a fixed point, it should be fine to also populate them. As noted, I would expect this to also simplify fixed-length vectors so if the existing canonicalization ones also include fixed-length pattern it should be ok to also have them here.

MacDue commented:
It means value bounds analysis would be needlessly run on CreateMask operations that from simple inspection are not all-true masks.

Say you had `create_mask %dynamic, %c2 : vector<4x[4]xi1>`. There's no canonicalization that will change that, however it's obvious from simply pattern matching the second dim that this won't be an all-true mask. So by checking these cases while we look for unknown dims we can bail out _before_ doing value bounds analysis. This check is like what the canonicalizations do, but actually using the canonicalizations won't give an early exit for obviously not all-true cases.

MacDue committed:
Review fixups

banach-space commented on mlir/test/Dialect/Vector/eliminate-masks.mlir:
```diff
@@ -0,0 +1,142 @@
+  %c4_vscale = arith.muli %vscale, %c4 : index
+  // Since %c1 is a constant, this will be found not to be all-true via simple
+  // pattern matching.
+  %mask = vector.create_mask %c1, %c4_vscale : vector<2x[4]xi1>
```
With `%c2` or `%c3` this would work, right? I'd test that as well.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorMaskElimination.cpp:
```diff
@@ -0,0 +1,131 @@
+    if (failed(dimLowerBoundSize))
+      return failure();
+    if (dimLowerBoundSize->scalable) {
+      // If the lower bound is scalable and < the mask dim size then this dim is
+      // not all-true.
+      if (dimLowerBoundSize->baseSize < maskTypeDimSizes[i])
+        return failure();
+    } else {
+      // If the lower bound is a constant:
+      // - If the mask dim size is scalable then this dim is not all-true.
+      if (maskTypeDimScalableFlags[i])
+        return failure();
+      // - If the lower bound is < the _fixed-size_ mask dim size then this dim
+      // is not all-true.
```
```suggestion
      // 1. The lower bound, LB, is scalable. If LB < the mask dim size then this dim is
      // not all-true.
      if (dimLowerBoundSize->baseSize < maskTypeDimSizes[i])
        return failure();
    } else {
      // 2. The lower bound, LB, is a constant.
      // 2.1 If the mask dim size is scalable then this dim is not all-true.
      if (maskTypeDimScalableFlags[i])
        return failure();
      // 2.2 If LB < the _fixed-size_ mask dim size then this dim
      // is not all-true.
```

[nit] This way I find easier to see what the distinct cases are.

banach-space commented:
> * Could we move the trivial pattern-matching cases to the folder/canonicalizer?

+1 IIUC, Diego is asking for more code re-use. In particular, if similar logic is already available somewhere in folder/conanicalization "landscape", it should be re-used here:
```cpp
struct UnknownMaskDim {
    size_t position;
    Value dimSize;
};

// Return `Failure` if at least one Mask dim is statically known not to be "all true". Otherwise,
// returns a vector of `UnknownMaskDim` yet to be analysed to see whether these are "all-true" lanes.
// If the return value is an empty vector, all Mask dims are statically known to be "all-true".
FailureOr<SmallVector<UnknownMaskDim>> getPossiblyTrueMaskDim() {
  SmallVector<UnknownMaskDim> unknownDims;
  for (auto [i, dimSize] : llvm::enumerate(createMaskOp.getOperands())) {
    if (auto intSize = getConstantIntValue(dimSize)) {
      // Mask not all-true for this dim.
      if (maskTypeDimScalableFlags[i] || intSize < maskTypeDimSizes[i])
        return failure();
    } else if (auto vscaleMultiplier = getConstantVscaleMultiplier(dimSize)) {
      // Mask not all-true for this dim.
      if (vscaleMultiplier < maskTypeDimSizes[i])
        return failure();
    } else {
      // Unknown (without further analysis).
      unknownDims.push_back(UnknownMaskDim{i, dimSize});
    }
  }
  return unknownDims;
}
```

From what I can tell, you can wrap this into an utilility funciton ,e.g. 
```cpp
SmallVector<>
```

> * Could we use `constant_mask` instead of `arith.constant`? This is not enforced in any way but we've been using the former to represent vector masks (they are kind of redundant, actually).

+1 to this suggestion. @dcaballe , are you OK to implement this as a follow-up?

MacDue commented:
> +1 IIUC, Diego is asking for more code re-use. In particular, if similar logic is already available somewhere in folder/conanicalization "landscape", it should be re-used here:

There's nowhere (other than here) that wants to check for not all-true create masks, and collect unknown dimensions :slightly_smiling_face:  I was just saying the logic is similar, but not really to a reusable extent imo.

banach-space commented:
> > +1 IIUC, Diego is asking for more code re-use. In particular, if similar logic is already available somewhere in folder/conanicalization "landscape", it should be re-used here:
> 
> There's nowhere (other than here) that wants to check for not all-true create masks, and collect unknown dimensions 🙂 I was just saying the logic is similar, but not really to a reusable extent imo.

Then you could still add a helper somewhere in Vector/Utils.

dcaballe commented:
What I'm suggesting here is:
  1. Move lightweight mask removal patterns to the op canonicalization. We want to remove trivial masks as part of the canonicalization. Create a `populateLightweightMaskRemovalPatterns` for them.
  2. Keep the heavyweight mask removal patterns in the pass and also add the patterns in `populateLightweightMaskRemovalPatterns` for completeness.

In that way, we would have the trivial ones as part of canonicalization and when more heavyweight analysis is needed to remove the masks, we will use the pass. Does it make more sense now or am I missing something?

MacDue commented:
> Does it make more sense now or am I missing something?

I think you are misunderstanding how this patch works. There are no lightweight canonicalizations added here (or any patterns really). Such lightweight canonicalizations already exist, and this code does not re-implement them, I was simply saying this code follows a similar idea.
 
Let's look at the actual code here:
```c++
  // Check for any dims that could be (partially) false before doing the more
  // expensive value bounds computations.
  SmallVector<UnknownMaskDim> unknownDims;
  for (auto [i, dimSize] : llvm::enumerate(createMaskOp.getOperands())) {
    if (auto intSize = getConstantIntValue(dimSize)) {
      // Mask not all-true for this dim.
      if (maskTypeDimScalableFlags[i] || intSize < maskTypeDimSizes[i])
        return failure();
    } else if (auto vscaleMultiplier = getConstantVscaleMultiplier(dimSize)) {
      // Mask not all-true for this dim.
      if (vscaleMultiplier < maskTypeDimSizes[i])
        return failure();
    } else {
      // Unknown (without further analysis).
      unknownDims.push_back(UnknownMaskDim{i, dimSize});
    }
  }
```
This is the part likened to canonicalization, but it's not a canonicalization. It's looking at all the operations of the `create_mask` and finding the unknown dimensions. Those are the dimensions that need to be solved via value-bounds analysis. The trick here is if any dimension is constant and not all-true, then we can exit early as that means the mask won't be all-true.

This prevents us from pointlessly doing value-bounds analysis is in cases like: 

```mlir
%mask = vector.create_mask %dynamicValue, %c2 : vector<8x4xi1>
```

From looking at `%c2` we can tell this is not going to be an all-true mask, so we don't need to run the value-bounds analysis for  
`%dynamicValue` (and will exit the transform early). Note that this `create_mask` still is not a `constant_mask` or an all-true/false mask, so no canonicalization will remove it before the "heavyweight" mask removal.

After that we go over the unknown dimensions and solve for them using value-bounds:
```c++
  for (auto [i, dimSize] : unknownDims) {
    // Compute the lower bound for the unknown dimension (i.e. the smallest
    // value it could be).
    FailureOr<ConstantOrScalableBound> dimLowerBound =
        vector::ScalableValueBoundsConstraintSet::computeScalableBound(
            dimSize, {}, vscaleRange.vscaleMin, vscaleRange.vscaleMax,
            presburger::BoundType::LB);
    if (failed(dimLowerBound))
      return failure();
    auto dimLowerBoundSize = dimLowerBound->getSize();
    if (failed(dimLowerBoundSize))
      return failure();
    if (dimLowerBoundSize->scalable) {
      // If the lower bound is scalable and < the mask dim size then this dim is
      // not all-true.
      if (dimLowerBoundSize->baseSize < maskTypeDimSizes[i])
        return failure();
    } else {
      // If the lower bound is a constant:
      // - If the mask dim size is scalable then this dim is not all-true.
      if (maskTypeDimScalableFlags[i])
        return failure();
      // - If the lower bound is < the _fixed-size_ mask dim size then this dim
      // is not all-true.
      if (dimLowerBoundSize->baseSize < maskTypeDimSizes[i])
        return failure();
    }
  }
```
 
The TLDR is this is just the "heavyweight" mask removal with a check to avoid doing pointless work.

banach-space reviewed (commented):
Sorry, forgot to press "send" before.

dcaballe commented:
Thanks for the clarifications!

> I think you are misunderstanding how this patch works. There are no lightweight canonicalizations added here (or any patterns really). Such lightweight canonicalizations already exist, and this code does not re-implement them, I was simply saying this code follows a similar idea.

Sorry if that is the case but I'm not sure how else I could interpret this:

> It attempts to do this by simply pattern-matching the mask operands (similar to some canonicalizations), if that does not lead to an answer (is all-true? yes/no), then value bounds analysis will be used to find the lower bound of the unknown operands.

The lightweight part that I'm referring as a canonicalization, based on this sentence, is the "pattern-matching the mask operands" that leads to an all-true answer without using the value bounds analysis.

More specifically: 

> Let's look at the actual code here:

My understanding is that if `unknownDims` is empty after that loop and no `failure()` path is taken, we are in an all-true mask case that do not require value bounds analysis (i.e., the lightweight case). If the answer to this is that this canonicalization already exists, then great! That's basically what I've been asking for :).

MacDue commented:
> My understanding is that if `unknownDims` is empty after that loop

That simply works as it would be more code to disable it. But there's already canonicalization of create_mask -> constant_mask that would handle those cases.

MacDue committed:
Fixups

MacDue commented:
I've now added a simple builder for making true/false constant_masks (after simplifying the representation in #100997.

```c++
  auto allTrue = rewriter.create<vector::ConstantMaskOp>(
      createMaskOp.getLoc(), maskType, ConstantMaskKind::AllTrue);
```

MacDue committed:
Share logic with `CreateMaskFolder`

The main thing shared here is the `getConstantVscaleMultiplier()`
matcher, I could not think of a good way to share all the logic as it's
somewhat different.

MacDue commented:
In the last commit, I've shared some logic with the `CreateMaskFolder`. The main thing shared here is the new `getConstantVscaleMultiplier()` matcher. I could not think of a good way to share all the logic, as it's a little different in both places.

banach-space reviewed (approved):
Thanks for all the refactor and code re-use. The split between `CreateMaskFolder` and `eliminateVectorMasks` is much clearer now. 

It would be good to briefly contrast what you are adding here with the current folder in the summary. That's where most people will be looking for documentation.

I believe that this addresses all of Diego's concerns. I know that he's OOO over the coming weeks and it would be really good to land this sooner rather than later. I will accept this (% my comments re documentation) and let's land this. If there's more suggestions from Diego post-commit, we can address those in follow-up PRs.

MacDue committed:
Update comments

MacDue commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -5862,73 +5887,46 @@ class CreateMaskFolder final : public OpRewritePattern<CreateMaskOp> {
+          return failure();
+        constantDims.push_back(*vscaleMultiplier);
+      } else {
+        return failure();
```
I've added a comments above both loops. Hopefully, the difference is clear now :slightly_smiling_face:

MacDue commented on mlir/lib/Dialect/Vector/Transforms/VectorMaskElimination.cpp:
```diff
@@ -0,0 +1,114 @@
+  };
+
+  // Check for any dims that could be (partially) false before doing the more
+  // expensive value bounds computations.
```
Updated comment (see above)
